tracks/elm/exercises/twelve-days/mentoring.md

```diff
@@ -0,0 +1,41 @@
+---
+
+This is a good exercise to talk about naming, as it is quite difficult and most students struggle with it.
+
+This exercise introduces some List handling functions, such as `List.map`, `List.range`, `List.take` and `List.reverse`, and some string handling functions, such as `++` and `String.join`.
+
+Students should strive to communicate what the code does in as few lines as possible (ideally you should be able to understand what is it trying to do in the entry function). They show also strive to communicate how it does it in as few lines as possible.
+
+### Reasonable solutions
+
+Simply listing out the complete text of all verses is the simplest and easiest solution, and would be a reasonable initial choice in real life. Then, if changes were required to the song, it would make sense to refactor to remove the duplication, and then make the changes. However this solution doesn't really help the student learn elm, so if they have done this it makes sense to encourage them to make a further iteration.
+
+All solutions need a way of turning an `Int` in to a string ordinal, and in to a string describing the gift for a day. `Dict.fromList`, `Array.fromList`, `List` or `case statements` are the normal ways to do this, and there is not much to choose between them.
+
+It is generally best to iterate over the verses using `List.map` or similar. It is possible to use recursion, but it makes naming harder, and is more complex than it needs to be.
+
+It is possible to use a type (like the one below) to contain all the information for a day (the ordinal and the gift for the day). This keeps the ordinal and gift strings closer together, which is good, but adds some noise, which is bad. Either is acceptable.
+
+```elm
+type alias NthDay =
+    { nthDay : Int
+	, ordinal : String
+    , gift : String
+    }
+```
+
+["Ceddlyburge" has a example solution](https://exercism.io/tracks/elm/exercises/twelve-days/solutions/7b614615e44f4ea6b6a0aa1b919054de) that uses case statements.
+	
+### Common suggestions
+
+Naming is difficult in this solution, and most students do not choose good names. Generally the name should be one level of abstraction up from what the code does (so it still means something, but you can change the implementation without having to change the name). `onTheNthDay` and `whatWasGiven` are reasonable choices for commonly used functions. `onTheNthDay` doesn't mention Christmas, so would allow the code to be changed for a different event ('on the first day of easter ...') and still make sense, while still being meaningful enough to be useful. 'onTheNth' is also possible, which would allow the code to change to 'on the second year of the new millenium' without having to change the name, and again the name is still meaningful enough to be useful.
```

Comment:
```suggestion
Naming is difficult, especially in this exercise, and most students often do not choose good names. Generally a good rule is to have the name be one-level-of-abstraction-up from what the code does, so that it still conveys a somewhat concrete meaning, but the implementation can change without having to change the name. `onTheNthDay` and `whatWasGiven` are reasonable choices for commonly used functions. `onTheNthDay` doesn't mention Christmas, so would allow the code to be changed for a different event ('on the first day of easter ...') and still make sense, while still being meaningful enough to be useful. 'onTheNth' is also possible, which would allow the code to change to 'on the second year of the new millenium' without having to change the name, and again the name is still meaningful enough to be useful.
```

See if you agree with these changes. I really like your explanation here.